File: Week_02/G20190343020023/n-ary-tree-preorder-traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""
class Solution:

    def __init__(self):
        self.res = []

    def preorder(self, root: 'Node') -> List[int]:
        self.helper(root)
        return self.res

    def helper(self, root):
        if root:
            self.res.append(root.val)
            for child in root.children:
                self.helper(child)

File: Week_02/G20190343020357/NAryTreePreorderTraversal.java
package week02.recursion.tree;

import java.util.ArrayList;
import java.util.List;

public class NAryTreePreorderTraversal {
    class Node {
        public int val;
        public List<Node> children;

        public Node() {
        }

        public Node(int _val) {
            val = _val;
        }

        public Node(int _val, List<Node> _children) {
            val = _val;
            children = _children;
        }
    }


    public List<Integer> preorder(Node root) {
        List<Integer> list = new ArrayList<Integer>();
        helper(root, list);
        return list;
    }

    public void helper(Node node, List<Integer> list) {
        if (node == null) return;
        list.add(node.val);
        for (Node child : node.children) {
            helper(child, list);
        }
    }

}
